File: zw_brain/domain/resource_kind.py
"""资源类型（物化形态）读路径归一 — 单一事实源（D53：只支持 库表/文件/API）。

`adapters/legacy/mappers/_normalize_resource_kind` 在**入库闸门**折叠（写侧）；本模块是**读路径**
防御兜底：存量库里仍残留 legacy `resource_kind='folder'/'url'/'link'` 的行（在归一化上线前导入），
任何把 stored kind 投影到消费面（发现卡 / 资源详情 / 搜索结果 / 交付分流）的读路径都必须经此折叠，
否则 folder/url 会泄漏到「资源类型」筛选、徽标与详情分型（违 D53 + R12 裸英文）。
"""

from __future__ import annotations

from typing import Any

# legacy 同义词 → 收敛后的物化形态（D53）。文件夹/链接退役并入文件；service 是 API 的历史别名。
_READ_KIND_FOLD = {"folder": "file", "url": "file", "link": "file", "service": "api"}
_CANONICAL_KINDS = frozenset({"table", "file", "api"})


def canonical_resource_kind(raw: Any) -> str | None:
    """折叠任意 stored resource_kind 到 库表/文件/API；空/未知→None（读路径不臆造 table，诚实空态）。

    - folder/url/link → file，service → api，table/file/api 原样；
    - None / 空串 / 未知 token → None（消费面不渲染徽标/筛选项，绝不裸出 legacy 英文）。
    """
    if raw is None:
        return None
    text = str(raw).strip().lower()
    if not text:
        return None
    text = _READ_KIND_FOLD.get(text, text)
    return text if text in _CANONICAL_KINDS else None

File: zw-brain-web/src/lib/resourceKind.ts
/**
 * 资源类型（物化形态）中文标签 — 前端单一事实源（R-019③）。
 *
 * 此前 4 处各自维护 resource_kind→中文 映射且已分叉（ResourceCard service:'服务'、
 * providerProjection service:'接口'、P2CatalogDetail 无 service 键、P2Discovery 同样缺）。
 * 收口到此，全仓统一。
 *
 * 口径权威源 = 后端 zw_brain/domain/resource_kind.py（读路径折叠）
 * + zw_brain/domain/serializers/typed_resource_detail.py KIND_LABELS。
 * D53②：资源类型只支持「库表 / 文件 / API」——文件夹/链接退役（folder/url/link→file）；
 * service 是 API 的历史别名（service→api），统一显示「接口」（与后端 KIND_LABELS 一致，
 * 修正 ResourceCard 旧「服务」分叉）。
 */

/** legacy 同义词 → 收敛后的物化形态（镜像后端 _READ_KIND_FOLD）。 */
const KIND_FOLD: Readonly<Record<string, string>> = {
  folder: 'file',
  url: 'file',
  link: 'file',
  service: 'api',
};

const CANONICAL_KINDS = new Set(['table', 'file', 'api']);

/** 收敛后物化形态 → 中文标签（镜像后端 KIND_LABELS 的 canonical 子集）。 */
const KIND_LABELS: Readonly<Record<string, string>> = {
  table: '库表',
  file: '文件',
  api: '接口',
};

/** 折叠任意 stored resource_kind 到 table/file/api；空/未知 → null（不臆造）。 */
export function canonicalResourceKind(raw: unknown): string | null {
  if (raw === null || raw === undefined) return null;
  const text = String(raw).trim().toLowerCase();
  if (!text) return null;
  const folded = KIND_FOLD[text] ?? text;
  return CANONICAL_KINDS.has(folded) ? folded : null;
}

/**
 * resource_kind → 中文标签（单一事实源）。
 * 未知/空 → 空串（消费面不渲染徽标，不裸出 legacy 英文，守 R12）。
 */
export function resourceKindLabel(raw: unknown): string {
  const kind = canonicalResourceKind(raw);
  return kind ? KIND_LABELS[kind] : '';
}

/** 仅 canonical 三型的标签表（供筛选 chip 等需要遍历键的场景）。 */
export const CANONICAL_KIND_LABELS = KIND_LABELS;
